capsules/extra/src/nmea_i2c.rs

```diff
@@ -0,0 +1,183 @@
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::ErrorCode;
+
+// We can only read a small-ish number of bytes at a time
+// otherwise the GNSS device will crash and hold the I2C
+// clock low. 32 is the maximum we can read
+pub const I2C_BUFFER_LEN: usize = 24;
+
+pub const NMEA_BUFFER_LEN: usize = 128;
+
+pub struct I2cNmea<'a, I: I2CDevice> {
+    sentence_buffer: TakeCell<'static, [u8]>,
+    i2c_buffer: TakeCell<'static, [u8]>,
+    nmea_offset: Cell<usize>,
+    i2c: &'a I,
+    client: OptionalCell<&'a dyn NmeaClient>,
+}
+
+impl<'a, I: I2CDevice> I2cNmea<'a, I> {
+    pub fn new(i2c: &'a I, i2c_buffer: &'static mut [u8]) -> Self {
+        I2cNmea {
+            sentence_buffer: TakeCell::empty(),
+            i2c_buffer: TakeCell::new(i2c_buffer),
+            nmea_offset: Cell::new(0),
+            i2c,
+            client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<'a, I: I2CDevice> NmeaDevice<'a> for I2cNmea<'a, I> {
+    fn set_client(&self, client: &'a dyn NmeaClient) {
+        self.client.set(client);
+    }
+
+    fn read_sentence(
+        &self,
+        sentence: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        match self.i2c_buffer.take() {
+            Some(buffer) => {
+                let i2c_buf_len = buffer.len();
+
+                match self.i2c.read(buffer, i2c_buf_len) {
+                    Ok(()) => {
+                        self.sentence_buffer.replace(sentence);
+                        Ok(())
+                    }
+                    Err((e, buf)) => {
+                        self.i2c_buffer.replace(buf);
+                        Err((e.into(), sentence))
+                    }
+                }
+            }
+            None => Err((ErrorCode::BUSY, sentence)),
+        }
+    }
+}
+
+impl<'a, I: I2CDevice> I2CClient for I2cNmea<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        let nmea_buf = self.sentence_buffer.take().unwrap();
+        let i2c_buf_len = buffer.len();
+
+        if let Err(e) = status {
+            self.i2c_buffer.replace(buffer);
+
+            self.client.map(|call| {
+                call.callback(nmea_buf, 0, Err(e.into()));
+            });
+
+            return;
+        }
+
+        let string = match str::from_utf8(buffer) {
+            Ok(utf8) => utf8,
+            Err(_e) => {
+                self.i2c_buffer.replace(buffer);
+
+                self.client.map(|call| {
+                    call.callback(nmea_buf, 0, Err(ErrorCode::NOSUPPORT));
+                });
+                return;
+            }
+        };
+        let mut nmea_offset = self.nmea_offset.get();
+
+        if let Some(location) = string.find('$') {
+            // This is the start of a new sentence
+
+            if (nmea_offset + location) > nmea_buf.len() {
+                // We will overflow our buffer, just drop the packet and try again
```

Comment:
Okay, my general problem here is that this sentence-copying state machine is pretty hard to follow and doesn't really seem all that correct. I _think_ it could be much simpler (and therefore harder to mess up), depending on how the I2C reads actually work.

Is it possible for I2C reads to give you part of an old sentence and part of a new sentence? For example, if it's possible to read partially mixed sentences you could have a sequence of reads like: the first read is "$oldold", the second read is "oldold$newnew", and the third read is "newnew".


### If you can read partially mixed sentences

In this case, the sentence looks something like "oldold$newnew". And the "oldold" doesn't fit in the sentence buffer, so you drop it, which seems reasonable.

However, shouldn't this try to copy from the "$" onwards to get the start of the next sentence? Otherwise, you've also dropped the "$newnew" part.

In general, if mixing sentences is the case, you might _always_ miss messages because it doesn't seem like there's a case here anywhere to _start_ reading a sentence partway through. And if you can mix messages, there's no guarantee that _any_ message will end up starting with a "$".

So in this case, a more clear design would be something like:
1. Determine if `nmea_offset` is zero
2. If `nmea_offset == 0`
  a. Skip message unless there's a $ in it
  b. Create a slice that's $ and onwards from i2c buffer
  c. Copy from slice into sentence and update offset
2. If `nmea_offset != 0`
  a. Create a slice that's start until $ (if any) from i2c buffer
  b. Copy from slice into sentence and update offset
  c. If there was a $ in the message, perform upcall


### If you cannot read partially mixed sentences

Then the only cases are 1) starts with a "$" or 2) does not.

A more clear design would be something like:

``` Rust
if string.starts_with('$') {
  if nmea_offset == 0 { 
    // start of sentence, copy into sentence buffer
  } else {
    // start of new sentence, upcall existing buffer and ignore
  }
} else {
  if nmea_offset == 0 {
    // ignore
  } else {
    // continuing sentence, copy into sentence buffer
  }
```

Reply:
> However, shouldn't this try to copy from the "$" onwards to get the start of the next sentence? Otherwise, you've also dropped the "$newnew" part.

True. If the `oldold` overflows our buffer we will incorrectly drop the `$newnew` part. I have never seen an overflow, so I didn't hit this code path. I'll fix it up

> In general, if mixing sentences is the case, you might _always_ miss messages because it doesn't seem like there's a case here anywhere to _start_ reading a sentence partway through. And if you can mix messages, there's no guarantee that _any_ message will end up starting with a "$".

We do handle reading a sentence mid way through, with this code

```rust
                        // The sentence didn't start with `$`. This usually occurs
                        // if we start reading mid-sentence. So we just try again and
                        // get the next one.
                        nmea_offset = 0;

                        // Copy the rest of the data
                        let size = i2c_buf_len - location;
                        nmea_buf[nmea_offset..(nmea_offset + size)]
                            .copy_from_slice(&buffer[location..]);
                        nmea_offset += size;
```

That gets hit if we don't run out of buffer space but the current sentence doesn't start with a `$`

Reply:
I just combed through the current iteration of the state machine, and I'm convinced it's correct.